lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -928,6 +931,98 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
   }
 }
 
+static GenericSignature *getNewGenericSignature(ValueDecl *base,
+                                                ValueDecl *derived) {
+  auto baseGenericCtx = base->getAsGenericContext();
+  auto &ctx = base->getASTContext();
+
+  if (!baseGenericCtx) {
+    return nullptr;
+  }
+
+  auto baseClass = base->getDeclContext()->getSelfClassDecl();
+
+  if (!baseClass) {
+    return nullptr;
+  }
+
+  auto derivedClass = derived->getDeclContext()->getSelfClassDecl();
+  auto *baseClassSig = baseClass->getGenericSignature();
+
+  if (!derivedClass) {
+    return nullptr;
+  }
+
+  if (derivedClass->getSuperclass().isNull()) {
+    return nullptr;
+  }
+
+  auto subMap = derivedClass->getSuperclass()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  if (baseGenericCtx->getGenericSignature() != nullptr) {
+    unsigned depth = 0;
+
+    if (auto *genericSig = baseClass->getGenericSignature())
+      depth = genericSig->getGenericParams().back()->getDepth() + 1;
+
+    GenericSignatureBuilder builder(ctx);
+    builder.addGenericSignature(derivedClass->getGenericSignature());
+
+    if (auto derivedGenericCtx = derived->getAsGenericContext()) {
+      if (derivedGenericCtx->isGeneric()) {
+        for (auto param : *derivedGenericCtx->getGenericParams()) {
+          builder.addGenericParameter(param);
+        }
+      }
+    }
+
+    auto source =
+        GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+
+    unsigned superclassDepth = 0;
+
+    if (baseClassSig) {
+      superclassDepth = baseClassSig->getGenericParams().back()->getDepth() + 1;
+    }
+
+    auto substFn = [&](SubstitutableType *type) -> Type {
+      auto *gp = cast<GenericTypeParamType>(type);
+
+      if (gp->getDepth() < superclassDepth) {
+        return Type(gp).subst(subMap);
+      }
+
+      return CanGenericTypeParamType::get(
+          gp->getDepth() - superclassDepth + depth, gp->getIndex(), ctx);
+    };
+
+    auto lookupConformanceFn =
+        [&](CanType depTy, Type substTy,
+            ProtocolDecl *proto) -> Optional<ProtocolConformanceRef> {
+      if (auto conf = subMap.lookupConformance(depTy, proto))
+        return conf;
+
+      return ProtocolConformanceRef(proto);
+    };
+
+    for (auto reqt : baseGenericCtx->getGenericSignature()->getRequirements()) {
+      if (auto substReqt = reqt.subst(substFn, lookupConformanceFn)) {
+        builder.addRequirement(*substReqt, source, nullptr);
+      }
+    }
+
+    if (builder.getGenericParams().empty()) {
```

Comment:
This is easy to determine before creating the GSB -- if derivedClass->getGenericSignature() is null and !baseGenericCtx->isGeneric(), there won't be any generic parameters.